lib/Sema/TypeCheckStmt.cpp

```diff
@@ -2519,6 +2526,20 @@ bool TypeCheckASTNodeAtLocRequest::evaluate(
       if (isa<TapExpr>(E))
         return Action::SkipChildren(E);
 
+      // If the location is within a single-expression branch of a
+      // SingleValueStmtExpr, don't walk into it. We'll want to type-check it
+      // a part of the whole expression.
+      if (auto *SVE = dyn_cast<SingleValueStmtExpr>(E)) {
+        SmallVector<Expr *> scratch;
+        for (auto *branch : SVE->getSingleExprBranches(scratch)) {
+          auto branchCharRange = Lexer::getCharSourceRangeFromSourceRange(
+            SM, branch->getSourceRange());
+          if (branchCharRange.contains(Loc))
+            return Action::SkipChildren(E);
+        }
```

Comment:
One case we could still improve here is if the there's an inner closure/SVE within the single-expression branch, as we could still potentially walk in and find something to type-check independently. I'm not sure such cases come up all that often tho.

Reply:
Actually I think it should be simple enough to handle, let me update.